### ISSUE TITLE
Can't access any methods of GDScript classes (except .new()) unless you first cast them as Script or GDScript

### ISSUE THREAD

**Issue by NotDaze** [godotengine/godot#76414]
### Godot version

4.0.2

### System information

Windows 11

### Issue description

You can't use any methods of a class (except .new()) unless you explicitly cast it to Script or GDScript.

```
class Example:
	pass;

func _init():
	print(Example is GDScript); # true
	print((Example as GDScript).has_source_code()); # false
	print((Example as Script).has_source_code()); # false
	print(Example.has_source_code()); # error
	
	var typed_implicit := Example; # also doesn't work
	print(typed_implicit.has_source_code()); # error
```

There is also some other... strangeness going on.

```
func _init():
	var untyped = Example;
	
	# print(untyped.has_source_code()); # error
	# print(Example.has_source_code); # error
	print(untyped.has_source_code); # works??
	print(untyped.has_source_code.call()); # also works??
	print(Example["has_source_code"].call()); # also works
```

Likely related to #73140

### Steps to reproduce

Attempt to access the methods of a GDScript-based class without first casting it to Script or GDScript.

### Minimal reproduction project

[MRP.zip](https://github.com/godotengine/godot/files/11314125/MRP.zip)

**Comment by YuriSizov** [comment]
`Example` in your example is neither `Script` nor `GDScript`, it's `RefCounted` (or `Reference` in 3.x).

```
print(Example is GDScript); # true
```
seems like an error to me, since it's not even a script, it's an inner class.

**Comment by NotDaze** [comment]
Everything is still the same if Example is a class_name in a separate file. Also, `Example.new().get_script()` returns a GDScript object. It appears that inner class objects themselves do extend GDScript.

**Comment by YuriSizov** [comment]
No, they don't. And if they somehow do, it's an error. If you don't specify any base type, `RefCounted` is used. But there is trickery to make it script-based types work that expose some of those `GDScript` methods, like `new()`.

**Comment by NotDaze** [comment]
The object you get from Example.new() does not implement GDScript. Example itself _absolutely_ does. I don't know if that's an error, but it is 100% what I am seeing.

**Comment by NotDaze** [comment]
Ok I'm actually partly wrong. `Example` itself is a RefCounted, but the fact that you can coerce it into a GDScript/Script object despite that is very weird and probably the cause of a lot of this strangeness.

**Comment by YuriSizov** [comment]
Yes, you are correct, of course. I got confused.
